### ISSUE TITLE
Option for image URI instead of base64

### ISSUE THREAD

**Issue by benydc** [cordova-plugin-camera-preview/cordova-plugin-camera-preview#294]
If I get to code an option for image native URI both android and ios, will you accept a pull request?

The base64 return method is memory intensive and lags 1,5 sec after taking the picture.

**Comment by westonganger** [comment]
Doesn't that assume that the photo is stored to file? We had so many bugs and problems related to file storage so we instead chose base64. Let me know how you plan to go about this and how it works.

**Comment by benydc** [comment]
We can either do it nativeURI and store it in the gallery/photoalbum or fileURI and save it in temp file path.

**Comment by westonganger** [comment]
Hmm... Well we ripped this feature out because it was causing so many problems. So im quite skeptical. If its really necessary and you can implement it in a simple and maintainable manner then I might be willing. 

One major problem area was permissions breaking with different OS versions. Possibly there is another (well maintained) cordova plugin you can make use of to abstract the permissions away. Also are you willing to help maintain this portion of the code if the functionality is merged?

**Comment by benydc** [comment]
Yes, I am looking at other cordova plugins that use the save to file and I will check as well regarding permission but shouldn't be a problem. Also, there is functionality to clear the temp dir at the control of the developer if we need it. If I get it work properly, yes I will maintain it.

**Comment by SVANNER** [comment]
Hi, It should be nice to get this!

Should it be possible to put an option to get the image the same as the size of the preview ?
The major added value of the plugin is to get a preview inside the app and directly how the image wil appear in the app. I currently cannot get the same aspects of the picture shoot as it appears in the preview. I tried with cropping the base64 image, but for the time being, i did not succeed.
I am ready to share anything that can help in this thing, also if you need testing on both Android and IOS, do not hesitate !

**Comment by westonganger** [comment]
@SVANNER there should be another issue for this or create another one. We used to have that functionality (on android at least) but we replaced it with the native resolution. Didn't know it would come with all this grief. If you want to take a stab at implementing it, the commit where we removed it was: 406e0c833cf0d2f2dd846b4840d929809260d9e6. Although you seem to be a wizard and might have your own solution. The final implementation you should be able to specify it in the takePicture options. Something like `type: "native" `or `"type: preview"` maybe?

**Comment by hug963** [comment]
I think the main issue with the first implementation is that it mixed everything.
These are two different things, one stops the preview, takes a proper picture and restarts the preview.
The other one just extracts a frame from the preview.

I think this should be put in a separate method than `takePicture` if you implement it. something like `getPreviewFrame`.

Also regarding the other issue of saving directly on the device, you can have a look in the [Cordova Camera](https://github.com/apache/cordova-plugin-camera) plugin, they implement saving to a temp file and separately saving to the photoalbum.

Saving to a temp file comes with all kind of drawbacks like having a separate method for iOS to clear the temp directory that needs to be called manually, dealing with permissions, etc and is a bit of a nightmare.

On the other hand, saving to photo Album seems relatively simple to implement:
https://github.com/apache/cordova-plugin-camera/blob/master/src/ios/CDVCamera.m#L502-L505
https://github.com/apache/cordova-plugin-camera/blob/master/src/android/CameraLauncher.java#L495-L506

**Comment by benydc** [comment]
Exactly! I have seen this on the old version. I reverted back to 0.0.8 cause it's cleaner and faster. There is no permission to be handled on the temp directory because it's inside the app and gets cleared by OS, but I agree that will need cleaning manually or inside the native, is pretty easy from the cordova-plugin-camera as example. At the moment it works wonders on iOS, extract from frame, save to temp all done instantly and will do android in the next days. 

The thing with saving to album, that should be a separate plugin that handles permission to access the users photo album... Whenever the user decides to save a photo, he will then be asked for the permissions on image picker ;-)

**Comment by hug963** [comment]
Hum, I think the new version is cleaner and faster! 

iOS actually doesn't clear the temp cache automatically and they have a `cleanup()` method that you need to call manually. https://github.com/apache/cordova-plugin-camera#module_camera.cleanup

It is not really faster as you need to move that file and clear the cache later...

I'm wondering why you are getting such big delays with the base64 versions, are you loading 100% quality images? I get full frames 90% images and it doesn't lag that much even when testing on my old iPhone 4S?

**Comment by benydc** [comment]
I had big delays when using the takePicture with "stops the preview, takes a proper picture and restarts the preview."

Reason why I switched to file is because I don't want to return a base64 and do domsanitizer to allow unsafe for img src. 

iOS does clear the temp, obviously not after app restarts but the OS might:
https://cordova.apache.org/docs/en/latest/reference/cordova-plugin-file/
Can be easily done as automatic and I am customizing this plugin for our apps need.

**Comment by JohanRonstrom** [comment]
Sorry for possible OT, but the "something like `getPreviewFrame`" that you mention, is that currently possible? To get the current preview frame as base64, without stopping the preview?

**Comment by benydc** [comment]
@JohanRonstrom yes it is, it's what I am using, it's instantly fast but quality is not good. 
I am building another plugin entirely in swift 3 and it's absolutely amazing. It's fast and high quality base64. for iOS of course for the moment.
